### PR TITLE
Add ADR for checking node configuration files with cardano-cli

### PR DESCRIPTION
**Pull request by smelc**
ADR explaining how we came up with the CLI's `check-node-configuration` command.

I've left the various commits to allow reviewers to see the decision's history, but I will squash before merging.

### CODE REVIEW

File: docs/ADR-5-cardano-testnet-node-configuration-file.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
+# Status
+
+📜 Proposed 2024-09-10
+
+# Context
+
+There is no canonical way to create a node configuration file. `cardano-cli` offers this possibility as part of the bigger [create-cardano](https://github.com/IntersectMBO/cardano-cli/blob/551d9b9f2f244e0d681bf03eaa6d985565ac3a5b/cardano-cli/test/cardano-cli-golden/files/golden/help/latest_genesis_create-cardano.cli#L49) command, but this is ad-hoc.
```
Comment (carbolymer) [inline docs/ADR-5-cardano-testnet-node-configuration-file.md:7]: Can you highlight more status quo here?

File: docs/ADR-5-cardano-testnet-node-configuration-file.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+Programmatic users proceed as follows instead:
+
+* Scripts like [mkfiles.sh](https://github.com/IntersectMBO/cardano-node/blob/51a034a51c5cefdd6ab4b9ff1e71710cf0c96643/scripts/babbage/mkfiles.sh#L85) and [jobs.nix](https://github.com/input-output-hk/cardano-parts/blob/0abf510e0ed70fda4e6ad3ae71632c20d09f135a/flakeModules/jobs.nix#L253) copy an existing template.
+* [cardano-testnet](https://github.com/IntersectMBO/cardano-node/blob/51a034a51c5cefdd6ab4b9ff1e71710cf0c96643/cardano-testnet/src/Testnet/Defaults.hs#L340) uses a hardcoded Haskell value.
+
+# Decision
```
Comment (carbolymer) [inline docs/ADR-5-cardano-testnet-node-configuration-file.md:14]: Were any other alternatives considered here?

**Review by Jimbo4350 — COMMENTED** [review]
I don't think we should have a command to  create a node configuration file as we will be repeating ourselves in a sense (why not just create the config file to begin with?). On the other hand a command that checks a node config file for the user would be handy. For example:
- The files listed do exist (and can be successfully decoded?) 
- The mandatory configuration values are set

File: docs/ADR-5-cardano-testnet-node-configuration-file.md [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+## Add a new command in the CLI
+
+Introduce a new `conway genesis check-node-config file` command in `cardano-cli`. It will have the following options:
+
+```
```
Comment (Jimbo4350) [inline docs/ADR-5-cardano-testnet-node-configuration-file.md:20]: Why list the options again? It should just check the existence of filepaths and that the genesis file hashes are correct. Have a look at the other fields that it may make sense to have checks for as well but to re-enter the file paths doesn't make sense imo. Parameterizing it on the era makes sense because things can change era to era and notifying the user what they may be missing in an upcoming era would be useful.

File: docs/ADR-5-cardano-testnet-node-configuration-file.md [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+[--shelley-genesis FILEPATH]
+[--alonzo-genesis FILEPATH]
+[--conway-genesis FILEPATH]
+[--node-config FILEPATH]
+```
+For every `--era-genesis` file, the command will read the file specified at the given `FILEPATH`, hash its content and check that the file specified at `--node-config` contains the correct path and the correct hash for this genesis file.
+
+## Have `create-testnet-data` create the node configuration file
+
+1. Have `create-testnet-data` create a default node configuration file (populating it with the paths and hashes of the genesis files).
```
Comment (Jimbo4350) [inline docs/ADR-5-cardano-testnet-node-configuration-file.md:31]: :+1:

File: docs/ADR-5-cardano-testnet-node-configuration-file.md [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+2. Add a `--node-config-template` optional option to `create-testnet-data which will be used to generate the node configuration file, instead of using a default one.
```
Comment (Jimbo4350) [inline docs/ADR-5-cardano-testnet-node-configuration-file.md:32]: This is problematic because the options exposed by `create-testnet-data` have to do with the modification of the genesis files _not_ the node configuration yaml file (besides perhaps the specification of where those genesis files live). Because of this I don't see much point in specifying an "override" option here.

File: docs/ADR-5-cardano-testnet-node-configuration-file.md (hunk 3 above)
Comment (smelc) [reply docs/ADR-5-cardano-testnet-node-configuration-file.md:20]: Yes, actually this was an earlier design and the actual WIP work ([this branch](https://github.com/IntersectMBO/cardano-cli/tree/smelc/check-node-config)) has the following API:

```
cardano-cli debug check-node-configuration
  --node-configuration-file FILEPATH
  [--fix-configuration-file]
```

Updating the ADR accordingly right now :+1:

File: docs/ADR-5-cardano-testnet-node-configuration-file.md (hunk 5 above)
Comment (smelc) [reply docs/ADR-5-cardano-testnet-node-configuration-file.md:32]: Sounds good, changing :+1:

File: docs/ADR-5-cardano-testnet-node-configuration-file.md (hunk 1 above)
Comment (smelc) [reply docs/ADR-5-cardano-testnet-node-configuration-file.md]: In the end I didn't, because the scope changed: we only check the configuration file, we don't create it.

File: docs/ADR-5-cardano-testnet-node-configuration-file.md [hunk 6]
```diff
@@ -0,0 +1,42 @@
+# Status
+
+📜 Proposed 2024-09-10
+
+# Context
+
+There is no way to check a node configuration file for basic sanity issues.
```
Comment (carbolymer) [inline docs/ADR-5-cardano-testnet-node-configuration-file.md:7]: Can you describe the risks and who is affected by not checking the configuration file?
I see that you described it in the *Consequences*, but I think it belongs more to the *Context*.

File: docs/ADR-5-cardano-testnet-node-configuration-file.md [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+# Decision
+
+## Add a new command in the CLI
+
+Introduce a new `debug check-node-configuration FILEPATH` command in `cardano-cli`. It will have the following options:
+
+```
+--node-configuration-file FILEPATH
```
Comment (carbolymer) [inline docs/ADR-5-cardano-testnet-node-configuration-file.md:16]: I would avoid listing the command names and option names. Those are parts of the user interface design, and not really n part of an architecture. Additionally, the commands and option names can change during the time and the ADR will be out of sync.

If you insist on keeping the command and option names, can you rephrase the text to make them examples rather than a references to a command option / toggle?
For example instead of

>### Have a `--fix` flag
>
>We considered having a `--fix` flag that would make `check-node-configuration`:

write

>### Provide fix up capability
>
>We considered having a dedicated flag (e.g. `--fix`) that would make the command:

File: docs/ADR-5-cardano-testnet-node-configuration-file.md (hunk 5 above)
Comment (smelc) [reply docs/ADR-5-cardano-testnet-node-configuration-file.md]: Resolving because deprecated

File: docs/ADR-5-cardano-testnet-node-configuration-file.md (hunk 6 above)
Comment (smelc) [reply docs/ADR-5-cardano-testnet-node-configuration-file.md:7]: Good point :+1: 

Adding more details in the _Context_ section:

```diff
 # Context
 
-There is no way to check a node configuration file for basic sanity issues.
+There is no way to check a node configuration file for basic sanity issues. This is a problem for people running testnets, because they will only discover _after having started their testnet_ that something is wrong (because the node will refuse to start). And starting a testnet is a costly operation, meaning the failures will only happen after tenth of seconds; causing the feedback loop to be slow.
```

File: docs/ADR-5-cardano-testnet-node-configuration-file.md (hunk 7 above)
Comment (smelc) [reply docs/ADR-5-cardano-testnet-node-configuration-file.md]: @carbolymer> good point :+1:

I addressed your concerns in the additional commit `Do not give exact API in the ADR. Make it more high-level.`

**Review by carbolymer — APPROVED** [review]
I'm ok with this. 👍🏻 

I remember there was a brief discussion about parsing of node config files, where the datatypes should lie and what's the difference between the cardano-node parsing of config file and what's done in api/cli. If there were conclusions, they would be worth adding here I think.

**Comment by smelc** [comment]
> I remember there was a brief discussion about parsing of node config files, where the datatypes should lie and what's the difference between the cardano-node parsing of config file and what's done in api/cli. If there were conclusions, they would be worth adding here I think.

@carbolymer> Made more explicit the move of types that was envisioned and what the types do in the additional commit `Make more explicit the move of types that was envisioned` :+1:
